### PR TITLE
Add fail_fast option for APRProofs

### PR DESCRIPTION
**Pull request by nwatson22**
Adds an option to abort proof execution early, leaving remaining nodes pending, when a node is detected that meets the following condition:
1. It is terminal (based on the `is_terminal` predicate
2. It is failing (failed subsumption into target).

### CODE REVIEW

File: src/pyk/proof/reachability.py [hunk 1]
```diff
@@ -609,14 +609,20 @@ def advance_pending_node(
         cut_point_rules: Iterable[str] = (),
         terminal_rules: Iterable[str] = (),
         implication_every_block: bool = True,
-    ) -> None:
+        fail_fast: bool = False,
+    ) -> bool:
```
Comment (tothtamas28) [inline src/pyk/proof/reachability.py:613]: What does the return value stand for?

Comment (nwatson22) [reply src/pyk/proof/reachability.py]: It was just a flag to tell the calling function to break out of the proof, but that was messy, so I figured out a way where that is unnecessary now.

File: src/pyk/proof/reachability.py [hunk 2]
```diff
@@ -655,6 +656,15 @@ def advance_proof(
                 terminal_rules=terminal_rules,
                 implication_every_block=implication_every_block,
             )
+            if (
+                fail_fast
+                and self.kcfg_explore.kcfg_semantics.is_terminal(curr_node.cterm)
+                and (not self.proof.kcfg.is_covered(curr_node.id))
```
Comment (tothtamas28) [inline src/pyk/proof/reachability.py:662]:
```suggestion
                and not self.proof.kcfg.is_covered(curr_node.id)
```

**Review by tothtamas28 — APPROVED** [review]

File: src/pyk/proof/reachability.py (hunk 2 above)
Comment (nwatson22) [reply src/pyk/proof/reachability.py]: Done.

[the remark this answers is not in the document]
Comment (nwatson22) [reply src/pyk/proof/reachability.py]: Done.

File: src/pyk/proof/reachability.py [hunk 3]
```diff
@@ -655,6 +656,15 @@ def advance_proof(
                 terminal_rules=terminal_rules,
                 implication_every_block=implication_every_block,
             )
+            if (
+                fail_fast
+                and not self.proof.kcfg.is_covered(curr_node.id)
+                and self.kcfg_explore.kcfg_semantics.is_terminal(curr_node.cterm)
+            ):
```
Comment (ehildenb) [inline src/pyk/proof/reachability.py:663]: @nwatson22 can't we just say `if (fail_fast and self.proof.status == ProofStatus.FAILING):` and leave it up to the proof type to determine whether it's failing or not?

**Pull request merged**